src/evaluators/DoWhileStatement.js

```diff
@@ -33,34 +34,67 @@ export default function(
   let V = realm.intrinsics.undefined;
 
   // 2. Repeat
-  while (true) {
-    // a. Let stmt be the result of evaluating Statement.
-    let stmt = env.evaluateCompletion(body, strictCode);
-    invariant(stmt instanceof Value || stmt instanceof AbruptCompletion);
+  let resultOrDiagnostic = realm.evaluateWithUndoForDiagnostic(() => {
```

Comment:
What does diagnostic mean in this context?

Reply:
If prepack is unable to interpret a loop because of an unknown loop bound, or because of some other problem in the loop body, it will issue a diagnostic and throw a FatalError. This helper function captures the diagnostic, squashes the exception and undoes any effects from the abortive loop interpretation, leaving the field open for the fixed point computation to have a go.